### ISSUE TITLE
Test coverage tracking

### ISSUE THREAD

**Issue by Mecharyry** [novoda/novoda#51]
## Issue
What we use to track test coverage seems to be a bit up in the air on our repositories. We have used `Sonar` for `Android` and `Groovy` projects in the past. Some have experimented with `CodeCov` too. This issue is a discussion on:

- What tracking do we currently use? Advantages and disadvantages of these?
- Do we currently know how to implement these? Is it documented for new projects?

## Potential Solution
Try to come up with a template for NOS projects so that we can have an easy to setup baseline for new projects.

**Comment by Mecharyry** [comment]
>  It depends on the outcome. We will want to present the coverage data somehow which means we need to push it somewhere

I would assume that we should provide the business with a report that shows the test coverage for all of our projects. @blundell might have some ideas about what he would like to see.

> we could create an all in one solution (gradle plugin) that generates the report and then uploads the result to a given url or we can force every project to use sonar since that already has a an api we can query.

I think it depends on the amount of work that is required for each. We would need to get some time allotted for this. Would the portfolio team put this sort of thing in their backlog @blundell ?

**Comment by mr-archano** [comment]
A Gradle plugin to collect information for a Java/Android project should be pretty easy to implement as the build chain integrates already with tools like JaCoCo outputting detailed reports (both HMTL and XML). I'd love to see a `novoda-pulse` plugin that we can apply to every project to collect important information automatically (test coverage, methods count, target API, etc)

**Comment by lgvalle** [comment]
Sonar is already collecting information about test coverage, won't be easier to just grab it from there? Is that possible?

**Comment by Mecharyry** [comment]
@lgvalle the issue atm is a) not all projects use sonar b) some require special configs which might mean maintaining a fork 😬  I'll update in here based on the Slack discussion.

**Comment by mr-archano** [comment]
In my experience Sonar API is bit clunky, and not fully supported. Also we would need to do some work anyway to extract info and mangle them in a pulse-compliant format. As already mentioned Gradle builds already output reports that are way easier to elaborate, we just need to find out where to send the collected data.

**Comment by ouchadam** [comment]
yep sonar has an api and we're already querying it for the legacy dashboards. 

I'd personally like to avoid sonar as it's a layer on top of a jacoco (for java) set up anyways.

**Comment by lgvalle** [comment]
Ok I see. Then, as @Mecharyry , mentioned in the description I think we first need to clarify:
1. What are we doing for our projects today?
2. What else do we want to achieve that we can't with the tools we already have?
3. What are the possible solutions to investigate

**Comment by blundell** [comment]
All projects should be running Sonar already (the original problem was lack of swift support)

I'd give Sonar another chance @ouchadam its been some time :)
